### PR TITLE
feat(dropdown): align, position, offset props + automatic position

### PR DESCRIPTION
**Pull request by bmdalex**
## BREAKING CHANGES MITIGATION
Replace `Dropdown` variables:

- ~~`borderRadius`~~ :arrow_right: `containerBorderRadius`
- ~~`openBorderRadius`~~ :arrow_right: `openAboveContainerBorderRadius` and `openBelowContainerBorderRadius`
- ~~`listBorderRadius`~~ :arrow_right: `aboveListBorderRadius` and `belowListBorderRadius` 

#### This PR tackles:
- introducing `align`, `position`, `offset` props for `Dropdown` component
- automatic positioning for `Dropdown` results
- reuses code from #1153 to avoid double render and simplify code

Screen Recording 2019-05-10 at 18 05 46

#### TODO:
- [x] Refactor method names and clean code
- [x] Extract common code for positioning elements to `src/lib/positioner`
- [x] Integrate new `Positioner` helper component to `Popup` and `Dropdown`
- [x] Refactor unit tests for `positionerHelper` (move from `Popup-test.tsx` to dedicated file)
- [x] Figure out way how to add some extra padding between `Dropdown` trigger and list of results
**any ideas?**

### CODE REVIEW

File: packages/react-component-ref/src/Ref.tsx [hunk 1]
```diff
@@ -14,7 +14,7 @@ export interface RefProps {
    *
    * @param {HTMLElement} node - Referred node.
    */
-  innerRef: React.Ref<any>
```
Comment (layershifter) [inline packages/react-component-ref/src/Ref.tsx]: This change is not correct. `innerRef` will be passed to `RefForwardRef` and there are no guarantees that it will be `HTMLElement`

File: docs/src/examples/components/Dropdown/Variations/DropdownExamplePosition.shorthand.tsx [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import * as React from 'react'
+import { Grid, Dropdown } from '@stardust-ui/react'
+
+const inputItems = ['Bruce Wayne', 'Natasha Romanoff', 'Steven Strange', 'Alfred Pennyworth']
+
+const DropdownArrowExample = props => {
+  const { position, align } = props
```
Comment (layershifter) [inline docs/src/examples/components/Dropdown/Variations/DropdownExamplePosition.shorthand.tsx]: Just wonder, will knobs more usable for this example?

File: packages/react/src/components/Dropdown/Dropdown.tsx [hunk 3]
```diff
@@ -236,6 +245,7 @@ class Dropdown extends AutoControlledComponent<Extendable<DropdownProps>, Dropdo
       content: false,
     }),
     activeSelectedIndex: PropTypes.number,
+    align: PropTypes.oneOf(_.without(ALIGNMENTS)),
```
Comment (layershifter) [inline packages/react/src/components/Dropdown/Dropdown.tsx]:
```suggestion
    align: PropTypes.oneOf(ALIGNMENTS),
```

Cruft?

File: packages/react/src/lib/positioner/UpdatableComponent.tsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  /**
+   * Array of conditions to be met in order to trigger a subsequent render.
+   */
+  updateDependencies: any[]
+}
+
+const UpdatableComponent: React.FunctionComponent<Extendable<UpdatableListProps>> = props => {
+  const { Component, innerRef, scheduleUpdate, updateDependencies, ...rest } = props
+
+  React.useEffect(() => scheduleUpdate && scheduleUpdate(), updateDependencies)
```
Comment (layershifter) [inline packages/react/src/lib/positioner/UpdatableComponent.tsx]:
```suggestion
  React.useEffect(() => scheduleUpdate(), updateDependencies)
```

It defined as required in `UpdatableListProps`...

File: packages/react/src/lib/positioner/createPopperReferenceProxy.ts [hunk 5]
```diff
@@ -4,11 +4,7 @@ import * as React from 'react'
 import * as PopperJS from 'popper.js'
 
 class ReferenceProxy implements PopperJS.ReferenceObject {
-  ref: React.RefObject<HTMLElement>
-
-  constructor(refObject) {
-    this.ref = refObject
-  }
+  constructor(private ref: React.RefObject<HTMLElement>) {}
```
Comment (layershifter) [inline packages/react/src/lib/positioner/createPopperReferenceProxy.ts]: Nit, but I am not sure about this change as I see it's TS specific stuff

[Babel REPL](https://babeljs.io/repl/#?babili=false&browsers=last%208%20versions%2C%20safari%20%3E%208%2C%20firefox%20%3E%2023%2C%20chrome%20%3E%2024%2C%20opera%20%3E%2015%2C%20not%20ie%20%3C%2011%2C%20not%20ie_mob%20%3C%3D%2011&build=&builtIns=false&spec=false&loose=false&code_lz=MYGwhgzhAECC0G8BQwD2A7CAXATgV2C1RwAocBTAMwEpEBfJBlDbaMaAXmgHIxulmmVCHIA6EKgDmJMKLDUgA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=es2015%2Creact%2Ctypescript&prettier=true&targets=Node-8.9&version=7.4.4&externalPlugins=)

File: packages/react/src/components/Dropdown/Dropdown.tsx [hunk 6]
```diff
@@ -559,20 +574,47 @@ class Dropdown extends AutoControlledComponent<Extendable<DropdownProps>, Dropdo
...
+
+    return (
+      <Positioner
+        align={align}
+        position={position}
+        offset={offset}
+        rtl={rtl}
+        target={this.selectedItemsRef}
+        children={popperChildrenProps => (
+          <UpdatableComponent
```
Comment (layershifter) [inline packages/react/src/components/Dropdown/Dropdown.tsx]: I just wondered, why `UpdatableComponent` is not part of `Positioner`? For me it has much more sence

File: packages/react/src/lib/positioner/createPopperReferenceProxy.ts (hunk 5 above)
Comment (bmdalex) [reply packages/react/src/lib/positioner/createPopperReferenceProxy.ts]: well, aren't we writing TS code? I don't understand the Babel reference it contains JS code..

File: packages/react-component-ref/src/Ref.tsx (hunk 1 above)
Comment (bmdalex) [reply packages/react-component-ref/src/Ref.tsx]: oops, reverted, but the doc comment above is misleading 😕

File: packages/react/src/components/Dropdown/Dropdown.tsx (hunk 3 above)
Comment (bmdalex) [reply packages/react/src/components/Dropdown/Dropdown.tsx]: oops, reverted, thanks

File: packages/react/src/components/Dropdown/Dropdown.tsx (hunk 6 above)
Comment (bmdalex) [reply packages/react/src/components/Dropdown/Dropdown.tsx]: I was thinking the same, but wanted to allow `Positioner` to receive components that don't necessarily update based on condition.

Alternative would be to:
- make `UpdatableComponent` HOC so that we don't need to pass `Component` and rest of props as props of `UpdatableComponent`
- move  `UpdatableComponent` to `Positioner` and add `updateDependencies` prop to `Positioner` as well and make it optional; then, if `updateDependencies` is:
  - `falsy`: we don't use `UpdatableComponent` at all
  - `[]`: component is rendered only when mounted (hooks behavior)
  - `[condition1, condition2, ..., conditionN]`: when at least a `conditionX` is true component rerenders (hooks behavior)

I think it makes sense. What do you think, @layershifter ?

File: packages/react/src/lib/positioner/UpdatableComponent.tsx (hunk 4 above)
Comment (bmdalex) [reply packages/react/src/lib/positioner/UpdatableComponent.tsx]: good point, applied suggestion, thanks

File: docs/src/examples/components/Dropdown/Variations/DropdownExamplePosition.shorthand.tsx (hunk 2 above)
Comment (bmdalex) [reply docs/src/examples/components/Dropdown/Variations/DropdownExamplePosition.shorthand.tsx]: hmm.. we can but for Popup we don't do that for very similar example. I can make the change for both for consistency and use knobs , but which one do you recommend?

@layershifter

File: packages/react/src/lib/positioner/createPopperReferenceProxy.ts (hunk 5 above)
Comment (layershifter) [reply packages/react/src/lib/positioner/createPopperReferenceProxy.ts]: We're going to use Babel as compiler, if we can't compile this code with Babel - it's a 💣 I didn't check it with local compiler, but was not able to make it working in REPL

File: packages/react/src/lib/positioner/Positioner.tsx [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+   * and 'start' | 'end' respectively), then provided value for 'align' will be ignored and 'center' will be used instead.
+   */
+  position?: Position
+}
+
+interface PositionerProps extends PopperProps, PositionCommonProps {
+  /**
+   * Content for children using render props API
+   */
+  children: (props: PopperChildrenProps) => React.ReactNode
```
Comment (layershifter) [inline packages/react/src/lib/positioner/Positioner.tsx]: It's defined in `PopperProps`, do we need it?

File: packages/react/src/lib/positioner/Positioner.tsx [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+  /**
+   * rtl attribute for the component
+   */
+  rtl?: boolean
+
+  target?: HTMLElement | React.RefObject<HTMLElement>
```
Comment (layershifter) [inline packages/react/src/lib/positioner/Positioner.tsx]: Missing description?

File: packages/react/src/components/Dropdown/Dropdown.tsx (hunk 6 above)
Comment (layershifter) [reply packages/react/src/components/Dropdown/Dropdown.tsx]: I suggest to remove `UpdatableComponent` at because it doesn't contain anything meaningful and place update logic in `Positioner`:

```tsx
  const scheduleUpdate = React.useRef<PopperChildrenProps['scheduleUpdate']>(null)
  
  React.useEffect(() => {
    if (scheduleUpdate.current) scheduleUpdate.current()
  }, updateDependencies) // `updateDependencies` is [] by default

  return (
    <Popper
      children={(childrenProps) => {
        scheduleUpdate.current = childrenProps.scheduleUpdate
        
        return children(childrenProps)
      }}
   />
```

Comment (layershifter) [reply packages/react/src/components/Dropdown/Dropdown.tsx]: And may be `positioningDependencies` will be more accurate, not sure

File: docs/src/examples/components/Dropdown/Variations/DropdownExamplePosition.shorthand.tsx (hunk 2 above)
Comment (layershifter) [reply docs/src/examples/components/Dropdown/Variations/DropdownExamplePosition.shorthand.tsx]: I think that we can both, it will make them more understandable and compact I think

File: packages/react/src/lib/positioner/createPopperReferenceProxy.ts (hunk 5 above)
Comment (bmdalex) [reply packages/react/src/lib/positioner/createPopperReferenceProxy.ts]: even if that's the case the solution is not to avoid TS features but to:
- either make compilation work with whatever compiler we use (e.g.: for babel there are plugins such as: https://babeljs.io/docs/en/babel-plugin-transform-typescript)
- or mark somehow (tslint config, lint rules, etc) what TS features we cannot use in the project (if there is strong reason).

I already see this TS feature being used in:
https://github.com/stardust-ui/react/blob/5e76653d940cd037ef9fc17931da9f4e2ac8327a/packages/react/src/lib/accessibility/FocusHandling/FocusContainer.ts#L4 

### tl;dr
I think preventing devs from using language features because our tooling might not work is wrong approach (unless that language feature has perf issues, increases bundle size or introduces problems).

File: docs/src/examples/components/Dropdown/Variations/DropdownExamplePosition.shorthand.tsx (hunk 2 above)
Comment (bmdalex) [reply docs/src/examples/components/Dropdown/Variations/DropdownExamplePosition.shorthand.tsx]: that took longer than expected; reimplemented Dropdown and Popup examples related to position using select knobs

File: packages/react/src/components/Dropdown/Dropdown.tsx (hunk 6 above)
Comment (bmdalex) [reply packages/react/src/components/Dropdown/Dropdown.tsx]: makes sense, the only thing missing was assigning the ref, added that following your suggestions, thanks!

File: packages/react/src/lib/positioner/Positioner.tsx (hunk 7 above)
Comment (bmdalex) [reply packages/react/src/lib/positioner/Positioner.tsx]: cruft, removed, thanks

File: packages/react/src/lib/positioner/Positioner.tsx (hunk 8 above)
Comment (bmdalex) [reply packages/react/src/lib/positioner/Positioner.tsx]: yup, added, thanks

File: packages/docs-components/src/knobs/useSelectKnob.ts [hunk 9]
```diff
@@ -1,9 +1,9 @@
 import { UseKnobOptions } from './types'
 import useKnob from './useKnob'
 
-const useSelectKnob = (options: UseKnobOptions<string>) =>
-  useKnob<string>({
-    initialValue: '',
+const useSelectKnob = <T = string>(options: UseKnobOptions<T>) =>
+  useKnob<T>({
+    initialValue: undefined,
```
Comment (bmdalex) [inline packages/docs-components/src/knobs/useSelectKnob.ts]: @layershifter this should be safe, right? is this still needed or can I omit?

File: docs/src/examples/components/Popup/Variations/PopupExamplePosition.shorthand.tsx [hunk 10]
```diff
@@ -1,54 +1,91 @@
...
+
+  React.useEffect(() => setPositionBeforeOrAfter(isPositionBeforeOrAfter(position)), [position])
+
+  const buttonStyles = { padding: paddings[position][align], height: '38px', minWidth: '64px' }
+
+  return (
+    <Grid columns="1" variables={{ padding: '100px 0' }} styles={{ justifyItems: 'center' }}>
+      <Popup
+        pointing
+        open={open}
```
Comment (miroslavstastny) [inline docs/src/examples/components/Popup/Variations/PopupExamplePosition.shorthand.tsx]: This means that if the checkbox is not checked, it is not possible to open the popup by clicking the trigger.
I would use something like `open={open || undefined}`

File: docs/src/examples/components/Popup/Variations/PopupExamplePosition.shorthand.tsx [hunk 11]
```diff
@@ -1,54 +1,91 @@
...
+          values: ['start', 'center', 'end'],
+        },
+  )
+
+  React.useEffect(() => setPositionBeforeOrAfter(isPositionBeforeOrAfter(position)), [position])
+
+  const buttonStyles = { padding: paddings[position][align], height: '38px', minWidth: '64px' }
+
+  return (
+    <Grid columns="1" variables={{ padding: '100px 0' }} styles={{ justifyItems: 'center' }}>
```
Comment (miroslavstastny) [inline docs/src/examples/components/Popup/Variations/PopupExamplePosition.shorthand.tsx:19]: Is `Grid` the best component for 0D layout?

File: docs/src/examples/components/Popup/Variations/PopupExamplePosition.shorthand.tsx [hunk 12]
```diff
@@ -1,54 +1,91 @@
...
-          <p>
-            The popup is rendered {position} the trigger
-            <br />
-            aligned to the {align}.
-          </p>
-        ),
-      }}
-    />
+  const [position] = useSelectKnob<Position>({
+    name: 'position shorthand',
```
Comment (miroslavstastny) [inline docs/src/examples/components/Popup/Variations/PopupExamplePosition.shorthand.tsx]: What does the word 'shorthand' mean in knob titles?

**Review by miroslavstastny — COMMENTED** [review]
The dropdown is jumping during page scroll:

dropdown_scroll

File: docs/src/examples/components/Popup/Variations/PopupExamplePosition.shorthand.tsx (hunk 12 above)
Comment (bmdalex) [reply docs/src/examples/components/Popup/Variations/PopupExamplePosition.shorthand.tsx]: I used it just to have a different name from the one in `PopupExamplePosition.tsx`. For knobs the name acts as an id so if I use the same name I get an error:

`Knob with name "${knob.name}" has been already registered`

Open to new name ideas that are more relevant

@miroslavstastny @layershifter

Comment (layershifter) [reply docs/src/examples/components/Popup/Variations/PopupExamplePosition.shorthand.tsx]: Will check it...

File: packages/docs-components/src/knobs/useSelectKnob.ts (hunk 9 above)
Comment (layershifter) [reply packages/docs-components/src/knobs/useSelectKnob.ts]: As I know it was changed in #1290, so you can just pick this change on merge

File: docs/src/examples/components/Popup/Variations/PopupExamplePosition.shorthand.tsx (hunk 10 above)
Comment (layershifter) [reply docs/src/examples/components/Popup/Variations/PopupExamplePosition.shorthand.tsx]: We can also use `onOpenChange`, each knob contains a setter in its tuple: `const [open, setOpen] = useBooleanKnob()`

File: docs/src/examples/components/Popup/Variations/PopupExamplePosition.shorthand.tsx (hunk 12 above)
Comment (layershifter) [reply docs/src/examples/components/Popup/Variations/PopupExamplePosition.shorthand.tsx]: `packages/docs-components/src/knobs/KnobProvider.tsx` should be updated with this:

```jsx
  const registerKnob = (knob: KnobDefinition) => {
    setKnobs(prevKnobs => {
      if (process.env.NODE_ENV !== 'production') {
        if (prevKnobs[knob.name]) {
          throw new Error(`Knob with name "${knob.name}" has been already registered`)
        }
      }

      return { ...prevKnobs, [knob.name]: knob }
    })
  }
```

File: docs/src/examples/components/Popup/Variations/PopupExamplePosition.shorthand.tsx (hunk 11 above)
Comment (bmdalex) [reply docs/src/examples/components/Popup/Variations/PopupExamplePosition.shorthand.tsx:19]: I don't know :( I used it because it was used in https://github.com/stardust-ui/react/blob/5c2688a9715b3b5e3925cc8da59420eada40a5c7/docs/src/examples/components/Popup/Variations/PopupExampleOffset.shorthand.tsx#L15

File: docs/src/examples/components/Popup/Variations/PopupExamplePosition.shorthand.tsx (hunk 10 above)
Comment (bmdalex) [reply docs/src/examples/components/Popup/Variations/PopupExamplePosition.shorthand.tsx]: It'd be a nice thing to add. It worked for the `Popup` but didn't for the `Dropdown`. When I deselect the `open` checkbox, Downshift starts updating the current state of `open` flag before the `open` prop is changed from `true` to `false`.

Order of events is strange.. I debugged this with @layershifter all day and didn't reach a conclusion :(

Any suggestions @miroslavstastny ? I'd keep this as is until we have issues impacting users.

File: docs/src/examples/components/Popup/Variations/PopupExamplePosition.shorthand.tsx (hunk 12 above)
Comment (bmdalex) [reply docs/src/examples/components/Popup/Variations/PopupExamplePosition.shorthand.tsx]: @layershifter sent me a fix for this so I was able to use just the same name, fixed

**Comment by bmdalex** [comment]
> The dropdown is jumping during page scroll:

@miroslavstastny I am aware of that. I was debugging popper code and noticed they recreate the popper instance (responsible for calculating styles that need to be applied for positioning elements) almost (they use debounce) every time there is a scroll/resize event: https://github.com/FezVrasta/popper.js/blob/36cc6c66e55f716232166c9a5725d8d08c4ec338/packages/popper/src/utils/setupEventListeners.js#L38 . This is where the jump comes from.

I think this can cause perf issues especially when trying to:
- position complex elements (such as the list of results in `Dropdown`);
- position multiple elements in the same view.

The jump is quite visible in our docs because there is a popper instance for **every Dropdown example** even if they are closed. This is due to the fact that for `Dropdown` we render the list of results even if it's empty (accessibility reasons, correct me if I'm wrong @silviuavram @jurokapsiar ).

### MITIGATION:

I think I was able to mitigate this by adding **`eventsEnabled={open}`** prop to `Popper`, which means the scroll/resize events are attached only when the list of results is open for a `Dropdown`. **@miroslavstastny can you please test again?**

### ALTERNATIVE:

I'm starting to think the simpler solution for `Dropdown` would be just to use `position: 'absolute'` and `top/bottom` CSS properties to position the the list of results above or below the trigger.

#### PROS:
- no jump;
- no perf issues because we won't recalculate anything on subsequent renders/scroll events/resize events;
- no need to integrate with popper at all.

#### CONS:
- we'd be limiting the `Dropdown` positioning to `above` or `below` the trigger; @jurokapsiar @silviuavram  @layershifter would that be enough according to our specs? I didn't really see too many libraries aiming for more than this;
- no auto positioning: we will need a function to tell us if the list of results is in the viewport to achieve that.

**Comment by miroslavstastny** [comment]
> * no auto positioning: we will need a function to tell us if the list of results is in the viewport to achieve that.

https://react.semantic-ui.com/behaviors/visibility/

File: packages/docs-components/src/knobs/KnobProvider.tsx [hunk 13]
```diff
@@ -12,13 +12,14 @@ const KnobProvider: React.FunctionComponent<KnobProviderProps> = props => {
   const [knobs, setKnobs] = React.useState<KnobSet>({})
 
   const registerKnob = (knob: KnobDefinition) => {
-    if (process.env.NODE_ENV !== 'production') {
-      if (knobs[knob.name]) {
-        throw new Error(`Knob with name "${knob.name}" has been already registered`)
+    setKnobs(prevKnobs => {
```
Comment (bmdalex) [inline packages/docs-components/src/knobs/KnobProvider.tsx]: fix for the problem with using a different names for the same knob, thx @layershifter

File: packages/react/src/lib/positioner/createPopperReferenceProxy.ts (hunk 5 above)
Comment (bmdalex) [reply packages/react/src/lib/positioner/createPopperReferenceProxy.ts]: NA anymore, code was removed

**Comment by bmdalex** [comment]
> The dropdown is jumping during page scroll:
> 
> dropdown_scroll

this perf issue was fixed by #1358

File: packages/docs-components/src/knobs/KnobProvider.tsx (hunk 13 above)
Comment (miroslavstastny) [reply packages/docs-components/src/knobs/KnobProvider.tsx]: This breaks existing docsite pages with knobs (`Button`, `Menu`):
image

Comment (bmdalex) [reply packages/docs-components/src/knobs/KnobProvider.tsx]: reverted so that I get unblocked, thanks, @layershifter let's tackle this fix in separate PR as the one proposed introduces the issue @miroslavstastny is mentioning

File: docs/src/examples/components/Dropdown/Variations/DropdownExamplePosition.shorthand.tsx [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+        align={align}
+        position={position}
+      />
+    </Grid>
+  )
+}
+
+export default DropdownExamplePosition
+
+type PositionAndAlign =
```
Comment (layershifter) [inline docs/src/examples/components/Dropdown/Variations/DropdownExamplePosition.shorthand.tsx]: Just wonder, do we really need this type?

File: docs/src/examples/components/Popup/Variations/PopupExamplePosition.shorthand.tsx [hunk 15]
```diff
@@ -1,54 +1,90 @@
...
-        position={position}
-        align={align}
-        icon={icon}
-        padding={padding}
-        key={`${position}-${align}`}
-      />
-    ))}
-  </Grid>
-)
+const icons: { [key in Position]: string } = {
```
Comment (layershifter) [inline docs/src/examples/components/Popup/Variations/PopupExamplePosition.shorthand.tsx]:
```suggestion
const icons: Record<Position, string> = {
```

 _nit_

File: packages/react/src/components/Dropdown/Dropdown.tsx [hunk 16]
```diff
@@ -1273,6 +1296,15 @@ class Dropdown extends AutoControlledComponent<WithAsProp<DropdownProps>, Dropdo
...
   }
+
+  private getToggleIndicatorIconName = (): string => {
+    const { open } = this.state
+    const positionedAbove = this.props.position === 'above'
+
+    return (open && !positionedAbove) || (!open && positionedAbove)
+      ? 'stardust-arrow-up'
+      : 'stardust-arrow-down'
+  }
```
Comment (layershifter) [inline packages/react/src/components/Dropdown/Dropdown.tsx]: This behavior is very strange, native `select` doesn't do the same thing:

image

File: packages/react/src/lib/positioner/Popper.tsx [hunk 17]
```diff
@@ -49,6 +49,7 @@ const Popper: React.FunctionComponent<PopperProps> = props => {
       const popperHasScrollableParent = getScrollParent(contentRef.current) !== document.body
 
       const modifiers: PopperJS.Modifiers = _.merge(
+        { preventOverflow: { padding: 0 } },
```
Comment (layershifter) [inline packages/react/src/lib/positioner/Popper.tsx:52]: Should be this in modifiers in `Dropdown` component?

File: packages/react/src/themes/teams/components/Dropdown/dropdownStyles.ts [hunk 18]
```diff
@@ -144,17 +147,14 @@ const dropdownStyles: ComponentSlotStylesInput<DropdownPropsAndState, DropdownVa
 
   list: ({ props: p, variables: v }): ICSSInJSStyle => ({
     outline: 0,
-    position: 'absolute',
-    borderRadius: v.listBorderRadius,
+    borderRadius: getOpenBorderStyles(p.position === 'below' ? 'top' : 'bottom', v),
```
Comment (layershifter) [inline packages/react/src/themes/teams/components/Dropdown/dropdownStyles.ts]: Can we omit this function and write it more flat, but more obvious?

```js
...p.open && p.position === 'top' && ({ borderBottomRaduis: v.borderBottomRadius }),
...p.open && p.position === 'bottom' && ({ borderBottomRaduis: v.borderBottomRadius }),
// etc
```

It's much more readable, see ##1301.

File: docs/src/examples/components/Dropdown/Variations/DropdownExamplePosition.shorthand.tsx (hunk 14 above)
Comment (bmdalex) [reply docs/src/examples/components/Dropdown/Variations/DropdownExamplePosition.shorthand.tsx]: Don't needed, just for safety. Placed it at the end so that it doesn't distract user. Can be removed if you feel strongly about it @layershifter

Comment (layershifter) [reply docs/src/examples/components/Dropdown/Variations/DropdownExamplePosition.shorthand.tsx]: As it was mentioned before, example should extra simple and contain less code. So if we don't have reasons to have this type, let's remove it.

File: packages/react/src/components/Dropdown/Dropdown.tsx (hunk 16 above)
Comment (bmdalex) [reply packages/react/src/components/Dropdown/Dropdown.tsx]: removed logic for now, will create issue

File: packages/react/src/lib/positioner/Popper.tsx (hunk 17 above)
Comment (bmdalex) [reply packages/react/src/lib/positioner/Popper.tsx:52]: I think it's something we want for all our components

File: packages/react/src/components/Dropdown/Dropdown.tsx (hunk 16 above)
Comment (bmdalex) [reply packages/react/src/components/Dropdown/Dropdown.tsx]: One of the aria examples is broken

File: packages/react/src/themes/teams/components/Dropdown/dropdownStyles.ts (hunk 18 above)
Comment (bmdalex) [reply packages/react/src/themes/teams/components/Dropdown/dropdownStyles.ts]: done, thanks

File: docs/src/examples/components/Dropdown/Variations/DropdownExamplePosition.shorthand.tsx (hunk 14 above)
Comment (bmdalex) [reply docs/src/examples/components/Dropdown/Variations/DropdownExamplePosition.shorthand.tsx]: done, thanks

File: docs/src/examples/components/Popup/Variations/PopupExamplePosition.tsx [hunk 19]
```diff
@@ -1,57 +1,77 @@
...
-        position={position}
-        align={align}
-        icon={icon}
-        padding={padding}
-        key={`${position}-${align}`}
-      />
-    ))}
-  </Grid>
-)
+const icons: { [key in Position]: string } = {
```
Comment (layershifter) [inline docs/src/examples/components/Popup/Variations/PopupExamplePosition.tsx]:
```suggestion
const icons: Record<Position, string> = {
```
_nit_

Comment (bmdalex) [reply docs/src/examples/components/Popup/Variations/PopupExamplePosition.tsx]: done, thanks

**Review by layershifter — APPROVED** [review]

**Review by miroslavstastny — APPROVED** [review]
